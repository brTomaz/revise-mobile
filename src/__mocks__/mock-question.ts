export const mockQuestion = {
  id: 1,
  info:
    '<figure class="image"><img src="https://res.cloudinary.com/dg64nrwjl/image/upload/v1608203495/question_4_c08561bb4d.png" alt=""></figure>',
  statement:
    'Ao abordar o assunto “obesidade”, em uma seção de jornal, o autor',
  answers: [
    'defende o consumo liberado de açúcar.',
    'aponta a gordura como o grande vilão da saúde.',
    'demonstra acreditar que a obesidade não é preocupante.',
    'indica a necessidade de mais pesquisas sobre o assunto.',
    'enfatiza a redução de ingestão de calorias pelos americanos.'
  ],
  correctAnswer: 4
}
