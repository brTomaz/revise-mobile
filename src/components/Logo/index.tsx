import React from 'react'
import { Image } from 'react-native'

import logo from '@/assets/images/logo.png'
import logoFirstLetter from '@/assets/images/logo-first-letter.png'

type LogoProps = {
  minimal?: boolean
}

const Logo = ({ minimal = false }: LogoProps) =>
  minimal ? (
    <Image source={logoFirstLetter} testID="minimal-logo" />
  ) : (
    <Image source={logo} testID="logo" />
  )

export default Logo
