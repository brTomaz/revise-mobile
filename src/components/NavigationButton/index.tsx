import React, { ReactNode } from 'react'
import { useNavigation } from '@react-navigation/native'

import * as Styles from './styles'

type NavigationButtonProps = {
  text: string
  icon: ReactNode
  navigateTo: string
}

const NavigationButton = ({
  text,
  icon,
  navigateTo
}: NavigationButtonProps) => {
  const navigation = useNavigation()

  async function handleNavigation(): Promise<void> {
    navigation.navigate(navigateTo)
  }

  return (
    <Styles.NavigationButton onPress={handleNavigation}>
      <Styles.ButtonText>{text}</Styles.ButtonText>
      <Styles.IconContainer>{icon}</Styles.IconContainer>
    </Styles.NavigationButton>
  )
}

export default NavigationButton
