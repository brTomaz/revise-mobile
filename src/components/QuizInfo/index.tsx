import React from 'react'
import * as Styles from './styles'

type QuizInfoProps = {
  mainInfo: number
  topText: string
  bottomText: string
}

const QuizInfo = ({ mainInfo, topText, bottomText }: QuizInfoProps) => (
  <Styles.Wrapper>
    <Styles.Text>{topText}</Styles.Text>
    <Styles.MainInfo>{mainInfo}</Styles.MainInfo>
    <Styles.Text>{bottomText}</Styles.Text>
  </Styles.Wrapper>
)

export default QuizInfo
