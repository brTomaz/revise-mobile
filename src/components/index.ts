export { default as Answer } from './Answer'
export { default as Logo } from './Logo'
export { default as NavigationButton } from './NavigationButton'
export { default as Question } from './Question'
export { default as QuizInfo } from './QuizInfo'
export { default as Statement } from './Statement'
