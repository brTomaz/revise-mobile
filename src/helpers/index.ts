export { default as configureTransition } from './transitions/configureTransition'
export { default as sleep } from './transitions/sleep'
export { default as getPoints } from './score/get-points'
export { default as shuffleQuestions } from './questions/shuffle-questions'
