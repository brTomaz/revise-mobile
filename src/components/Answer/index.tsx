import React, { useState } from 'react'

import { getPoints, showToast } from '@/helpers'
import { useQuiz } from '@/hooks/quiz'
import { State, AnswerProps } from './protocols'
import * as Styles from './styles'

const Answer = ({ text, answerIndex, correctAnswerIndex }: AnswerProps) => {
  const [state, setState] = useState(State.Unavailable)
  const {
    currentTries,
    points,
    setPoints,
    setCurrentTries,
    setCurrentQuestionIndex,
    currentQuestionIndex
  } = useQuiz()

  const updateQuizContext = (pointsEarned: number) => {
    setPoints(pointsEarned + points)
    setCurrentQuestionIndex(currentQuestionIndex + 1)
    setCurrentTries(0)
  }

  const handlePress = () => {
    setCurrentTries(currentTries + 1)

    const isWrongAnswer = answerIndex !== correctAnswerIndex

    if (isWrongAnswer) {
      setState(State.Wrong)
    } else {
      setState(State.Correct)
      const pointsEarned = getPoints(currentTries)

      showToast(
        'success',
        `Ganhou ${pointsEarned} ponto(s)`,
        `Acertou na ${currentTries + 1}Âª tentativa`
      )

      updateQuizContext(pointsEarned)
    }
  }

  return (
    <Styles.Wrapper
      onPress={handlePress}
      disabled={state !== State.Unavailable}
    >
      <Styles.Text state={state}>{text}</Styles.Text>
    </Styles.Wrapper>
  )
}

export default Answer
