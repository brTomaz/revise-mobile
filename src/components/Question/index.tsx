import React, { useState, useEffect } from 'react'
import { Animated } from 'react-native'

import Answer from '@/components/Answer'
import Statement from '@/components/Statement'
import { sleep, configureTransition } from '@/helpers'
import * as Styles from './styles'

type QuestionProps = {
  info: string
  statement: string
  answers: string[]
  correctAnswer: number
}

enum Transition {
  Starting = 'Starting',
  Completed = 'Completed'
}

const Question = ({
  info,
  statement,
  answers,
  correctAnswer
}: QuestionProps) => {
  const [transition, setTransition] = useState(Transition.Starting)

  const animatedOpacity = new Animated.Value(0)

  useEffect(() => {
    async function configTransition() {
      await sleep(500)
      await configureTransition(() => {
        setTransition(Transition.Completed)
      })

      Animated.timing(animatedOpacity, {
        toValue: 1,
        duration: 500,
        delay: 500,
        useNativeDriver: true
      }).start()
    }
    configTransition()
  })

  return (
    <Styles.Wrapper>
      {transition === Transition.Completed && (
        <Animated.View>
          <Styles.InfoWrapper>
            <Styles.Info source={{ html: info }} />
          </Styles.InfoWrapper>
          <Styles.AnswerAndStatementWrapper>
            <Styles.StatementWrapper>
              <Statement text={statement} />
            </Styles.StatementWrapper>
            {answers.map((answer, index) => (
              <Styles.AnswerWrapper key={answer}>
                <Answer
                  text={answer}
                  answerIndex={index}
                  correctAnswerIndex={correctAnswer - 1}
                />
              </Styles.AnswerWrapper>
            ))}
          </Styles.AnswerAndStatementWrapper>
        </Animated.View>
      )}
    </Styles.Wrapper>
  )
}

export default Question
