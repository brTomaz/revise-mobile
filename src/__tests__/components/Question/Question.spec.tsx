import React from 'react'

import { Question } from '@/components'
import { renderWithTheme } from '@/helpers'
import { mockQuestion } from '@/__mocks__/mock-question'

describe('<Question />', () => {
  jest.useFakeTimers('legacy')

  it('should match to snapshot', async () => {
    const { id, info, statement, answers, correctAnswer } = mockQuestion

    const { toJSON } = renderWithTheme(
      <Question
        key={id}
        info={info}
        statement={statement}
        answers={answers}
        correctAnswer={correctAnswer}
      />
    )

    expect(toJSON()).toMatchSnapshot()
  })
})
