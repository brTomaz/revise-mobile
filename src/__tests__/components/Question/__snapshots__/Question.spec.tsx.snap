// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Question /> should match to snapshot 1`] = `
<RCTScrollView
  style={
    Array [
      Object {},
    ]
  }
>
  <View>
    <View
      style={
        Object {
          "opacity": 0,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#FAFAFA",
              "borderBottomLeftRadius": 0,
              "borderRadius": 8,
              "justifyContent": "center",
              "marginBottom": 16,
              "paddingBottom": 12,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 12,
            },
          ]
        }
      >
        <View
          style={Object {}}
        >
          <View
            style={
              Array [
                Object {},
                Object {},
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignSelf": "stretch",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  Object {},
                ]
              }
            >
              <View
                style={
                  Object {
                    "height": 100,
                    "width": 100,
                  }
                }
                testID="image-placeholder"
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FAFAFA",
              "borderRadius": 8,
              "paddingBottom": 4,
              "paddingLeft": 4,
              "paddingRight": 4,
              "paddingTop": 4,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "marginBottom": 8,
                "paddingBottom": 0,
                "paddingLeft": 4,
                "paddingRight": 4,
                "paddingTop": 0,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#FAFAFA",
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#030617",
                    "fontFamily": "Quicksand-SemiBold",
                    "fontSize": 16,
                  },
                ]
              }
            >
              Ao abordar o assunto “obesidade”, em uma seção de jornal, o autor
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "marginTop": 4,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#040923",
                "borderRadius": 8,
                "opacity": 1,
                "paddingBottom": 8,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 8,
              }
            }
          >
            <Text
              state="Unavailable"
              style={
                Array [
                  Object {
                    "color": "#FAFAFA",
                    "fontFamily": "Quicksand-Regular",
                  },
                ]
              }
            >
              defende o consumo liberado de açúcar.
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "marginTop": 4,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#040923",
                "borderRadius": 8,
                "opacity": 1,
                "paddingBottom": 8,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 8,
              }
            }
          >
            <Text
              state="Unavailable"
              style={
                Array [
                  Object {
                    "color": "#FAFAFA",
                    "fontFamily": "Quicksand-Regular",
                  },
                ]
              }
            >
              aponta a gordura como o grande vilão da saúde.
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "marginTop": 4,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#040923",
                "borderRadius": 8,
                "opacity": 1,
                "paddingBottom": 8,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 8,
              }
            }
          >
            <Text
              state="Unavailable"
              style={
                Array [
                  Object {
                    "color": "#FAFAFA",
                    "fontFamily": "Quicksand-Regular",
                  },
                ]
              }
            >
              demonstra acreditar que a obesidade não é preocupante.
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "marginTop": 4,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#040923",
                "borderRadius": 8,
                "opacity": 1,
                "paddingBottom": 8,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 8,
              }
            }
          >
            <Text
              state="Unavailable"
              style={
                Array [
                  Object {
                    "color": "#FAFAFA",
                    "fontFamily": "Quicksand-Regular",
                  },
                ]
              }
            >
              indica a necessidade de mais pesquisas sobre o assunto.
            </Text>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "marginTop": 4,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#040923",
                "borderRadius": 8,
                "opacity": 1,
                "paddingBottom": 8,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 8,
              }
            }
          >
            <Text
              state="Unavailable"
              style={
                Array [
                  Object {
                    "color": "#FAFAFA",
                    "fontFamily": "Quicksand-Regular",
                  },
                ]
              }
            >
              enfatiza a redução de ingestão de calorias pelos americanos.
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
